examples/hosts/app-integration/external-data/src/model/taskList.ts

```diff
@@ -0,0 +1,190 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { DataObject, DataObjectFactory } from "@fluidframework/aqueduct";
+import { SharedCell } from "@fluidframework/cell";
+import { TypedEventEmitter } from "@fluidframework/common-utils";
+import { SharedString } from "@fluidframework/sequence";
+
+import { externalDataSource, parseStringData } from "../externalData";
+import type { ITask, ITaskEvents, ITaskList } from "../modelInterfaces";
+
+class Task extends TypedEventEmitter<ITaskEvents> implements ITask {
+    public get id() {
+        return this._id;
+    }
+    // Probably would be nice to not hand out the SharedString, but the CollaborativeInput expects it.
+    public get name() {
+        return this._name;
+    }
+    public get priority() {
+        const cellValue = this._priority.get();
+        if (cellValue === undefined) {
+            throw new Error("Expected a valid priority");
+        }
+        return cellValue;
+    }
+    public set priority(newValue: number) {
+        this._priority.set(newValue);
+    }
+    public constructor(
+        private readonly _id: string,
+        private readonly _name: SharedString,
+        private readonly _priority: SharedCell<number>,
+    ) {
+        super();
+        this._name.on("sequenceDelta", () => {
+            this.emit("nameChanged");
+        });
+        this._priority.on("valueChanged", () => {
+            this.emit("priorityChanged");
+        });
+    }
+}
+
+/**
+ * The TaskList is our data object that implements the ITaskList interface.
+ */
+export class TaskList extends DataObject implements ITaskList {
+    private readonly tasks = new Map<string, Task>();
+
+    public readonly addTask = (id: string, name: string, priority: number) => {
+        if (this.tasks.get(id) !== undefined) {
+            throw new Error("Task already exists");
+        }
+        const nameString = SharedString.create(this.runtime);
+        nameString.insertText(0, name);
+        const priorityCell: SharedCell<number> = SharedCell.create(this.runtime);
+        priorityCell.set(priority);
+        this.root.set(id, { id, name: nameString.handle, priority: priorityCell.handle });
```

Comment:
Shouldn't this be this.tasks.set?

Reply:
Good question, I'll add some comments in the code to clarify too --

Here I set the value on this.root to make the actual data change (and so other clients will see the new member of the root).  The this.tasks collection really just holds these local facades (Task class) on the data for individual items for convenience and encapsulation (flushes out the async so we can use a sync interface, give a convenient object to hand to a view rather than the whole this.root SharedMap).  So changing the this.tasks collection alone wouldn't result in any persisted data changes.

Instead, when this.root has a "valueChanged" event (which happens immediately when setting the new content on the root) we'll create and add the Task to the this.tasks collection (in handleTaskAdded).  Changing this.root triggers the change in this.tasks automatically.

Overall it's a slight variation on the collection pattern (see this demo for a more thorough dive on that pattern: https://github.com/microsoft/FluidFramework/tree/main/examples/apps/contact-collection).